Prioritised tax calculation strategies

mckelviemark opened the pull request:
Added support for prioritised tax calculation strategies. This includes a global taxation setting for the default tax calculation strategy. This includes two strategies, one being an order items based strategy and the other being order item units based. This work should make it easy for client projects to create and drop in custom strategies.

Commit:
- Create a single unified OrderTaxesApplicatorInterface for the tax applicators by zone to implement.
- Create a CompositeOrderTaxesCalculationStrategy to apply a hierarchical OrderTaxesApplicatorInterface.
- Create two separate tax calculation strategies, one that is order items based and the other being
  order item units based.
- Create a PrioritizedServiceRegistry and write a compiler pass to register the tax calculation strategies.
- Rename the OrderTaxesApplicator to OrderTaxesProcessor and inject the tax calculation strategies registry.
- Add a global default tax calculation strategy setting.

peteward commented on src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:
```diff
@@ -235,22 +238,62 @@
+                        factory-service="sylius.settings.manager"
+                        factory-method="loadSettings"
+                        class="Sylius\Bundle\SettingsBundle\Model\Settings"
+                >
```
I'm not sure about the multi-line split, seeing as they don't stick to 80 chars in their XML, but a small detail

peteward replied:
NOTE: the settings bundle has a huge pending PR and a drastic change is imminent...

mckelviemark replied:
ok, as for the formatting of the multi-line split this is how they have formatted it elsewhere, so I'll keep this as is to be consistent.

peteward commented on src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php:
```diff
@@ -26,15 +28,13 @@
 {
     /**
-     * Zone repository.
+     * Zone repository
```
you can just remove this comment

peteward commented on src/Sylius/Bundle/CoreBundle/Settings/TaxationSettingsSchema.php:
```diff
@@ -26,15 +28,13 @@
      * @var ObjectRepository
      */
-    private $zoneRepository;
+    protected $zoneRepository;
```
why change this?

mckelviemark replied:
I'll change it back. I was just trying to be consistent with the scope of their other member variables.

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php:
```diff
@@ -0,0 +1,94 @@
+/**
+ * @author Mark McKelvie <mark.mckelvie@reiss.com>
+ */
+class CompositeOrderTaxesCalculationStrategy implements OrderTaxesApplicatorInterface
```
should there be a common `TaxApplicatorInterface`? Is there any benefit or requirement for an `OrderTaxesApplicatorInterface` explicitly?

peteward replied:
ok ignore me, this is at the strategy level... Should a Strategy implement an Applicator interface? hmmm

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/CompositeOrderTaxesCalculationStrategy.php:
```diff
@@ -0,0 +1,94 @@
+     * @param Settings $settings
+     * @param OrderTaxesApplicatorInterface[] $applicators
+     */
+    public function __construct($type, Settings $settings, array $applicators)
```
does `$type` need to be injected? Isn't it dictated by the class? should this be a constant?

peteward replied:
Ignore this, I see what you're doing here.

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php:
```diff
@@ -0,0 +1,102 @@
+    {
+        foreach ($order->getItemUnits() as $unit) {
+            $item = $unit->getOrderItem();
+            $taxRate = $this->taxRateResolver->resolve($item->getVariant(), ['zone' => $zone]);
```
hmmm....

It might be more efficient to have 2 loops, first loops around orderItems so you can resolve the tax rate just once per `OrderItem`, then an inner loop which calculates and applies it for each unit in the item?

mckelviemark replied:
Good, I've updated this to calculate on every unit but to only resolve the tax rate once per item.

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php:
```diff
@@ -0,0 +1,102 @@
+     */
+    public function supports(OrderInterface $order, ZoneInterface $zone)
+    {
+        return false;
```
huh?

peteward replied:
OK again, I think perhaps a separation of the Strategy and Applicator interfaces means that you don't need `supports()` at the Applicator level - only Strategy.

gonzalovilaseca replied:
+1

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/OrderItemUnitsTaxesApplicator.php:
```diff
@@ -0,0 +1,102 @@
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Core\Taxation\Applicator;
```
newline after this

mckelviemark replied:
done.

peteward commented on src/Sylius/Component/Core/Taxation/Applicator/OrderTaxesApplicatorInterface.php:
```diff
@@ -9,21 +9,33 @@
+
+    /**
+     * @param OrderInterface $order
+     * @param ZoneInterface  $zone
```
sylius don't align comments

mckelviemark replied:
done.

peteward commented on src/Sylius/Bundle/TaxationBundle/Form/Type/TaxCalculationStrategyChoiceType.php:
```diff
@@ -0,0 +1,55 @@
+ */
+class TaxCalculationStrategyChoiceType extends AbstractType
+{
+    protected $strategies = [
```
any reason to be protected instead of private?

mckelviemark replied:
It's protected because this is the same way that Sylius does it for other ChoiceTypes. When I saw a similar pattern elsewhere on Sylius I thought that to get it working I would list the strategy types here. Of course this means I would have to extend it on Reiss and we can avoid that by injecting it into the constructor instead. That's what I'm going to try now.

peteward commented:
Mark, some really, really excellent work here.

Why have you made the PR into `mm-tax-applicator` instead of `master`? What other changes would be included in the PR?  I think I'd like to see the whole picture.

I only have one significant design issue and that's the Composite/Strategy structure. My _feeling_ is that we should move away from the labelling that the Strategies are CompositeApplicators. They are definitely a kind of wrapping layer around Applicators, but they are more than a standard Composite pattern and I think can benefit from a clearer separation. This is my _suggestion_, but we should discuss whether you do this now or we look for Sylius feedback first.
- Don't make the strategies implement an `XXXXApplicatorInterface`, make it a `TaxStrategyInterface`
- Establish some separation between Applicators and Strategies by differing the methods such as `applyTaxes()` on the `Strategy` and `apply` on the `Applicator`.
- You have one achilles heel in overriding which is the use of some hard-coded strings and the `TaxCalculationStrategies` class.
  - Firstly, I don't think this class is needed, because if you want to add a new strategy you either have to override/extend this class or your strategy just won't be listed.
  - Secondly I don't think it offers anything anyway. It's used to identify which strategies should be available through settings, but shouldn't you use the registry here? Otherwise it's not going to be truely flexible because you would have to override the form and add your hard-coded strategy option. So if you instead inject the registry and use this to get the name of strategies you have a double-win.

UPDATE: I've just seen why you called the strategy `Composite` because it's reusable for either OrderItem or OrderItemUnit calculations - clever. But I still think we should remove the work Composite because it's not (quite) the same pattern.

Have a read through of this and let me know your thoughts. I think the changes I'm suggesting are quite small so maybe it's worth doing them first if you agree?

I would definitely like to see the full PR too.

Commit:
- Code tidy and clean up.
- Separate out the tax calculation strategy from the tax applicators.
- Use tag attributes to create tax calculation strategies parameter that can then be injected into the TaxCalculationStrategyChoiceType service.
- Move the TaxCalculationStrategyChoiceType and TaxRateType.
- Throw an UnsupportedTaxCalculationStrategyException in the OrderTaxesProcessor.
- Update the order item units tax applicator to only resolve the tax rate once for each order item before performing the calculation on each order item unit.

mckelviemark commented:
@peteward The reason for the merge from mm-tax-applicator-review to mm-tax-applicator is that we already have a PR open with Sylius and everytime I push to that latter branch it triggers a build. Once I merge back to that branch we can add a PR comment to ping Sylius to review the approach. We'll need to point out that the PHPSpec is largely unfinished and possible Behat scenarios may need updating.

peteward commented:
You could just close your PR

mckelviemark commented:
You'll need to do that because it was you that opened it with my initial attempt. Anyway, we can merge this in very soon if you think it's fine for their review.

peteward commented on src/Sylius/Bundle/CoreBundle/Resources/config/services.xml:
```diff
@@ -235,22 +240,67 @@
+                <argument type="service" id="sylius.taxation.order_shipment_taxes_applicator" />
+            </argument>
+            <tag name="sylius.taxation.calculation_strategy" type="order_items_based" label="Order items based" />
+        </service>
```
newline here would be good

peteward commented on src/Sylius/Bundle/CurrencyBundle/Templating/Helper/CurrencyHelper.php:
```diff
@@ -58,13 +62,7 @@ public function convertAmount($amount, $currency = null)
-     * @param bool        $decimal
-     *
-     * @return string
+     * {@inheritdoc}
```
these changes look good but watch out for including unrelated changes in PRs, should be fine though as long as there's not many

mckelviemark commented:
I think I should probably rename `TaxCalculationStrategyInterface` to just `TaxStrategyInterface`. This will still make sense and it will allow me to shorten the name of `UnsupportedTaxCalculationStrategyException` to `UnsupportedTaxStrategyException`.

peteward commented:
I think it's fine as it is

mckelviemark commented:
ok. :-)
